plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +241,185 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+            Host host = _resourceMgr.addHost(zoneId, resource, Host.Type.L2Networking, params);
             if (host != null) {
-                return Transaction.execute(new TransactionCallback<NuageVspDeviceVO>() {
-                    @Override
-                    public NuageVspDeviceVO doInTransaction(TransactionStatus status) {
-                        NuageVspDeviceVO nuageVspDevice = new NuageVspDeviceVO(host.getId(), physicalNetworkId, ntwkSvcProvider.getProviderName(), deviceName);
-                        _nuageVspDao.persist(nuageVspDevice);
-
-                        DetailVO detail = new DetailVO(host.getId(), "nuagevspdeviceid", String.valueOf(nuageVspDevice.getId()));
-                        _hostDetailsDao.persist(detail);
-
-                        return nuageVspDevice;
-                    }
-                });
+                NuageVspDeviceVO nuageVspDevice = new NuageVspDeviceVO(host.getId(), physicalNetworkId, ntwkSvcProvider.getProviderName(), deviceName);
+                _nuageVspDao.persist(nuageVspDevice);
+
+                DetailVO detail = new DetailVO(host.getId(), "nuagevspdeviceid", String.valueOf(nuageVspDevice.getId()));
+                _hostDetailsDao.persist(detail);
+
+                ConfigurationVO cmsIdConfig = _configDao.findByName("nuagevsp.cms.id");
+                host = findNuageVspHost(nuageVspDevice.getHostId());
+                SyncNuageVspCmsIdCommand syncCmd = new SyncNuageVspCmsIdCommand(SyncType.REGISTER, null);
+                SyncNuageVspCmsIdAnswer answer = (SyncNuageVspCmsIdAnswer) _agentMgr.easySend(nuageVspDevice.getHostId(), syncCmd);
+                if (answer != null && answer.getSuccess()) {
+                    registerNewNuageVspDevice(cmsIdConfig, nuageVspDevice.getId() + ":" + answer.getNuageVspCmsId());
+
+                    detail = new DetailVO(host.getId(), "nuagevspcmsid", answer.getNuageVspCmsId());
+                    _hostDetailsDao.persist(detail);
+                    hostdetails.put("nuagevspcmsid", answer.getNuageVspCmsId());
+                    resource.configure((String) hostdetails.get("hostname"), hostdetails);
+
+                    auditDomainsOnVsp((HostVO) host, true, false);
+                } else {
+                    throw new CloudRuntimeException("Failed to register CMS ID");
+                }
+                return nuageVspDevice;
             } else {
                 throw new CloudRuntimeException("Failed to add Nuage Vsp Device due to internal error.");
             }
         } catch (ConfigurationException e) {
-            throw new CloudRuntimeException(e.getMessage());
+            s_logger.error("Failed to configure Nuage VSD resource", e);
+            throw new CloudRuntimeException("Failed to configure Nuage VSD resource", e);
+        } catch (ExecutionException ee) {
+            s_logger.error("Failed to add Nuage VSP device", ee);
+            throw new CloudRuntimeException("Failed to add Nuage VSP device", ee);
```

Comment:
Please add identifying information about the Nuage device to the error messages.